certora/ArrangerConduit.spec

```diff
@@ -0,0 +1,710 @@
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5 || revert6 ||
+                            revert7 , "Revert rules failed";
+
+}
+
+// Verify correct storage changes for non reverting withdraw
+rule withdraw(bytes32 ilk, address asset, uint256 maxAmount) {
+    env e;
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    mathint balanceOfBufferBefore = gem.balanceOf(buffer);
+    mathint balanceOfConduitBefore = gem.balanceOf(currentContract);
+    mathint withdrawableFundsBefore = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFundsBefore = totalWithdrawableFunds(asset);
+    mathint withdrawalsBefore = withdrawals(asset, ilk);
+    mathint totalWithdrawalsBefore = totalWithdrawals(asset);
+
+    require balanceOfBufferBefore + balanceOfConduitBefore <= max_uint256;
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset)); 
+    withdraw(e, ilk, asset, maxAmount);
+
+    mathint balanceOfBufferAfter = gem.balanceOf(buffer);
+    mathint balanceOfConduitAfter = gem.balanceOf(currentContract);
+    mathint withdrawableFundsAfter = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFundsAfter = totalWithdrawableFunds(asset);
+    mathint withdrawalsAfter= withdrawals(asset, ilk);
+    mathint totalWithdrawalsAfter = totalWithdrawals(asset);
+
+    assert balanceOfBufferAfter == balanceOfBufferBefore + amount, "balance of buffer did not increase by amount";
+    assert balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
+    assert withdrawableFundsAfter == withdrawableFundsBefore - amount, "withdrawableFunds did not decrease by amount";
+    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore- amount, "totalWithdrawableFunds did not decrease by amount";
+    assert withdrawalsAfter == withdrawalsBefore + amount, "withdrawals did not increase by amount";
+    assert totalWithdrawalsAfter == totalWithdrawalsBefore + amount, "totalWithdrawals did not increase by amount";
+}
+
+// Verify revert rules on withdraw
+rule withdraw_revert(bytes32 ilk, address asset, uint256 maxAmount) {
+    env e;
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xa6fb97d1)); // withdraw(bytes32,address,uint256)
+    mathint balanceOfConduit = gem.balanceOf(currentContract);
+    mathint withdrawals = withdrawals(asset, ilk);
+    mathint totalWithdrawals = totalWithdrawals(asset);
+    mathint withdrawableFunds = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFunds = totalWithdrawableFunds(asset);
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset)); 
+    withdraw@withrevert(e, ilk, asset, maxAmount);
+
+    bool revert1 = e.msg.value > 0;
```

Comment:
Same comment, weird we are not having a revert on balance of the destination.

Reply:
On the other hand, I have a personal preference to try to preserve the order of the reverts according to how they can happen in the code. Sometimes it is a bit difficult though as some revert rule might group different revert conditions, and also a bit annoying if you discovered one is missing in the beginning.

Reply:
nvm I just did it.